### PR TITLE
feat: metamask account migration

### PR DESCRIPTION
**Pull request by tomicvladan**
Metamask account migration dialog with notifications.

### CODE REVIEW

**Comment by IgorShadurin** [comment]
## Suggestions for migration process

1. **Metamask requests to sign the login text twice**: Why does migration ask Metamask to sign the same login text two times? If it's for obtaining the wallet, it can be retrieved without signing the login text (if user has logged in via Metamask).

2. **Migration modal closure on UI element click**: If I accidentally click on some UI element outside the migration modal, it will close. Could this cause issues if the user modifies some files or starts the migration process again during the previous cycle of migration? Perhaps the modal could be locked during the process?

3. **Display of network name during migration**: It would be helpful to show the name of the network during all steps of the migration. This ensures I am on the correct network. The network name could be displayed in a static location or mentioned in the text. 
As crypto operations require additional actions like mnemonic writing, adding networks, switching networks, topping up balances, etc., a user might not be certain about the exact network they chose on the previous step after returning to the UI.

    Screenshot 2023-07-21 at 10 57 59

4. **Migration modal window background**: Why doesn't the migration modal window have a background? It's not typical for modals.

    Screenshot 2023-07-21 at 11 00 03

5. **Info text display and complexity**: The information text can be displayed more compactly, like gray text with a smaller font size at the bottom of the modal. The list of transactions could be hidden in a list that appears by clicking the word about transactions. This would reduce the complexity of the form.

6. **No actions after clicking send button**: After clicking the send button (for the spinner or balance validation), there is no indication of whether the button was clicked or not. This could create confusion, especially in some systems/browsers where it can take time for the Metamask window to appear. Low priority.

7. **Early registration click**: What happens if the user clicks on Registration before receiving the balance? Can the migration process still proceed correctly?

8. **Balance discrepancy between Metamask and Fairdrive UI**: The minimum balance displayed in Metamask is less than that requested in the Fairdrive UI.
At least as displayed in Metamask, the value is less than what's shown on the UI. Maybe we should round the value on the UI to avoid showing too long a number?

    Screenshot 2023-07-21 at 11 00 54

9. **Gas estimation issue**: When I closed and reopened the modal, I received the message "Gas estimation can't be performed on this network." So, we should suggest a specific amount that would be enough for the respective network in most cases.

10. **Error when sending minimum balance after gas estimation issue**: After getting a message that "Gas estimation can't be performed on this network" and clicking the button to send the minimal balance, I encountered the error shown below.

    Screenshot 2023-07-21 at 11 20 29

11. **Transaction timeout error**: What is the timeout value? After receiving a timeout, I noticed some of the balance was spent (for `register` and `setResolver` actions). If I click Register again, there won't be enough balance to continue the process if starting it over. Also, my username will be occupied, right?

    Screenshot 2023-07-21 at 11 07 52


In total, our Create Account site operates more smoothly for these tasks. Perhaps we can reuse it? We could open the Create Account site with some parameters for migration, or even open it as an iframe. It handles a lot of things effectively.

**Review by IgorShadurin — CHANGES_REQUESTED** [review]
Left my question in the other comment

**Comment by tomicvladan** [comment]
1. Yes, it asks for signing the message to obtain mnemonic. Mnemonic could be saved in memory after logging in.
2. If the user closes the modal, that won't cause any issues. The only step where interaction with services is done is when "Register" button is clicked. But modal state is saved, so if the user opens the modal again, it will continue from last state.
3. Sure, just I'm not sure how it could be done from UI perspective. @tfius  any suggestions?
4. This is not a blocking modal, but it still can have darkened background.
5. In dark theme the text is a bit different, but it seems for light theme there is no difference. I'll fix that. I don't know if it's so important to put that much effort into small tooltip text.
6. OK, we can add a spinner while Metamask is being opened.
7. The "Register" button is not visible before initialization which includes geting account balance and estimating transactions price.
8. Metamask just truncates small decimals. The exact same amount it sent to Metamask and transaction is made with the same amount. It's just displayed differently.
9. I agree, but now when we have veriety of networks I'm not sure what value we should put for each.
10. This needs to be fixed, probably minBalance is null which is sent to Metamask.
11. That's right. Currently there is no way to reset registration from transaction that has failed. We need a registration retry mechanism in fdp-storage in order to do that. I think I created an issue for that there once.

Regarding the create account apporach. We could do something like that, but now when we already have this implementation, that might be a waste of time to repeat the same implementation there.

**Comment by tfius** [comment]
image
"Your FDS Lite account, managed with Metamask, can be seamlessly migrated to an FDS Portable account. This enables you to use a username and password for login on mobile devices and other dApps."

**Comment by tfius** [comment]
image
Text after `*info: ` should be as icon and tooltip should be displayed with text: 
Also `Register` button  should be either disabled or hidden if there is not enough balance. 
Rename it to `Continue registration`
So something like this; 

Your account is: 0x
Before proceeding with account registration, make sure you have enough funds on it.
Estimated minimal balance is 0.0000xxxx

<`Send Minimal balance`>
(check balance) once its ok
1. hide <`Send`> button
display
"**Funds sufficient**" 
<`Continue Registration`>

**Comment by tfius** [comment]
check for funds on account periodically !

**Comment by IgorShadurin** [comment]
I've recorded a video of the registration process. It's pretty long, but here are some notable points:
-The "Private" section of the drive isn't shadowed for some reason.
-I can't unblock the "Continue" button. At first, I thought the minimal balance might not be correct because even after sending it with the "Send" button, the "Continue Registration" button remained locked. I suspect the issue might be due to the balance retrieval process API, which occasionally returns errors. This could be blocking the "Continue" button, even though the button does get enabled briefly, but not long enough for a click.
In conclusion, I'm unable to complete the migration process.

Questions:
-What improvements could be made to make the migration process smoother for the user?
-What happens if a user chooses Sepolia and their Metamask is on Mainnet, and they click the "Send" button? Will it send real ETH to the address?


https://github.com/fairDataSociety/fairdrive-theapp/assets/1595061/87383fc5-2d7b-4f60-a7d8-eea0f7efdfe4

**Comment by tomicvladan** [comment]
This step is a bit complex because a lot is going on in the background. As a first step the modal checks what's minimal balance for registration. If that fails it shows a message saying that it couldn't get min balance and the continue button gets enabled. If it gets min balance, it shows it and immediately start checking current account balance periodically. If fetching balance works it disables the button if there is no min balance, otherwise it enables it. And now I made a fix, if getting current balance fails, it shows an error message and enables the button.
Also now before sending a transaction the app will try to switch to the selected network.
And regarding the highlighted button in the background, the reason for that is usage of z-index. So the z-index of the component is higher than the z-index of modal background. I'm not sure how to fix that because if I increase z-index of the background, then the modal becomes darkened. But applying z-index to the modal doesn't work.

**Comment by tomicvladan** [comment]
Do you know what's the purpose of z-index classes? It seems that they don't impact the look, so they can be deleted?

**Comment by IgorShadurin** [comment]
> Do you know what's the purpose of z-index classes? It seems that they don't impact the look, so they can be deleted?

Actually, I don't know what they are there for. I think if they can be removed without damaging the UI on PC and Mobile, then it's worth doing.

**Comment by tomicvladan** [comment]
I've removed z-index classes for all static elements and I didn't notice any issues.

**Comment by tfius** [comment]
we need to update docs with this process @tamas6
